app/lib/admin/backend.dart

```diff
@@ -0,0 +1,200 @@
+
+      // Package.uploaders
+      final pkgQuery = _db.query<Package>()..filter('uploaders =', user.userId);
+      await for (final p in pkgQuery.run()) {
+        await _removeUploaderFromPackage(p.key, user.userId);
+      }
+
+      // PublisherMember
+      // Publisher.contactEmail
+      final memberQuery = _db.query<PublisherMember>()
+        ..filter('userId =', user.userId);
+      await for (final m in memberQuery.run()) {
+        await _removeMember(user, m);
+      }
+
+      // OAuthUserID
+      if (user.oauthUserId != null) {
+        final key = _db.emptyKey.append(OAuthUserID, id: user.oauthUserId);
+        final mapping = (await _db.lookup<OAuthUserID>([key])).single;
+        if (mapping != null) {
+          await _db.commit(deletes: [mapping.key]);
+        }
+      }
+
+      // User
+      await _markUserDeleted(user);
+    });
+  }
+
+  Future _removeUploaderFromPackage(Key pkgKey, String userId) async {
+    await _db.withTransaction((tx) async {
+      final p = (await tx.lookup<Package>([pkgKey])).single;
+      p.removeUploader(userId);
+      if (p.uploaders.isEmpty) {
+        p.isDiscontinued = true;
+      }
+      tx.queueMutations(inserts: [p]);
+      await tx.commit();
+    });
+  }
+
+  Future _removeMember(User user, PublisherMember member) async {
+    final seniorMember =
+        await _remainingSeniorMember(member.publisherKey, member.userId);
+    await _db.withTransaction((tx) async {
+      final p = (await tx.lookup<Publisher>([member.publisherKey])).single;
+      if (seniorMember == null) {
+        p.contactEmail = null;
+      } else if (p.contactEmail == user.email) {
+        final seniorUser =
+            await accountBackend.lookupUserById(seniorMember.userId);
+        p.contactEmail = seniorUser.email;
+      }
+      tx.queueMutations(inserts: [p], deletes: [member.key]);
+      await tx.commit();
+    });
+  }
+
+  /// Returns the member of the publisher that (a) is not removed,
+  /// (b) preferably is an admin, and (c) is member of the publisher for the
+  /// longest time.
+  ///
+  /// If there are no more admins left, the "oldest" non-admin member is returned.
+  Future<PublisherMember> _remainingSeniorMember(
+      Key publisherKey, String excludeUserId) async {
+    final otherMembers = await _db
+        .query<PublisherMember>(ancestorKey: publisherKey)
+        .run()
+        .where((m) => m.userId != excludeUserId)
+        .toList();
+
+    if (otherMembers.isEmpty) return null;
+
+    // sort admins in the front, and on equal level sort by created time
+    otherMembers.sort((a, b) {
+      if (a.role == b.role) return a.created.compareTo(b.created);
+      if (a.role == PublisherMemberRole.admin) return -1;
+      if (b.role == PublisherMemberRole.admin) return 1;
+      return a.created.compareTo(b.created);
+    });
+
+    return otherMembers.first;
+  }
+
+  Future _markUserDeleted(User user) async {
+    await _db.withTransaction((tx) async {
```

Comment:
If the user doesn't have any interactions in `History` and doesn't have any interactions in `PackageVersion`, then we could delete the user, right?

But I guess we have no efficient way to query `History` entities to see if the user is present there.. Let's leave a TODO for this. I'm sure we can clean that up when we get around to refactoring the `History` entity.